### PR TITLE
Only call NotifyBlockTip when chainActive changes

### PR DESCRIPTION
**Pull request by jamesob**
This is a subset of the more controversial https://github.com/bitcoin/bitcoin/pull/12407, but this also adds a test demonstrating the bug.

In InvalidateBlock, we're calling NotifyBlockTip with the now-invalid block's prev regardless of what chain the ancestor block is on. This could create numerous issues, but it at least screws up `waitforblockheight` (or anything else relying on `rpc/blockchain.cpp:latestblock`) when InvalidateBlock is called on a block not in chainActive, which can happen via RPC.

Only call NotifyBlockTip when the block being marked invalid is on the active chain.

### CODE REVIEW

File: test/functional/rpc_blockchain.py [hunk 1]
```diff
@@ -227,6 +240,47 @@ def _test_stopatheight(self):
...
+        node.invalidateblock(b22f.hash)
+
+        def assert_waitforheight(height, timeout=2):
+            assert_equal(
+                node.waitforblockheight(height, timeout)['height'],
+                current_height)
+
+        assert_waitforheight(0)
+        assert_waitforheight(current_height - 3)
+        assert_waitforheight(current_height)
```
Comment (conscott) [inline test/functional/rpc_blockchain.py]: Might also want to add a timeout case like in `_test_stopatheight` above.

Comment (jamesob) [reply test/functional/rpc_blockchain.py]: Played around with doing this, but found that the return from RPC is (unfortunately) no different in the timeout case, meaning we'd have to spin up a separate thread in-test, run a timer, and then assert that the RPC thread hadn't yet stopped execution -- probably overkill.

Comment (jamesob) [reply test/functional/rpc_blockchain.py]: I've added a testcase for `current_height + 1`, though.

**Comment by jamesob** [comment]
Thanks for the look, @conscott. Rebased.

**Comment by conscott** [comment]
utACK f98b543522687a4ff93ad7d53fa3cc67b5c6d752

**Comment by ryanofsky** [comment]
@jamesob, I'm curious how this might interact with #11856 with gets rid of the uiInterface.NotifyBlockTip signal, replacing it with CValidationInterface::UpdatedBlockTip listeners.

**Review by ryanofsky — COMMENTED** [review]
utACK f98b543522687a4ff93ad7d53fa3cc67b5c6d752. This fixes a pretty clear `waitforblockheight` bug and adds a nice test case.

**Comment by sdaftuar** [comment]
utACK f98b543522687a4ff93ad7d53fa3cc67b5c6d752

**Comment by ryanofsky** [comment]
@TheBlueMatt, this is a simple, one line change fixing the `NotifyBlockTip` notification in `InvalidateBlock`. Could you confirm that it looks ok and doesn't mess up your plans for #11856?

**Comment by TheBlueMatt** [comment]
Without thinking about it too hard, this seems fine. I'd obviously like to move towards getting rid of uiInterface calls in validation, ala #11856, but this should likely be applied there as well.

**Comment by ryanofsky** [comment]
@laanwj, I think this looks good to merge

**Pull request merged**